### ISSUE TITLE
RAM consumption when loading models 

### ISSUE THREAD

**Issue by CoaNewco** [Unity-Technologies/barracuda-release#174]
Hello, we had some issues concerning RAM consumption. First of all, when we import the models in our project, automatically we see spike in our used memory. It seems that the models use double the memory as we are seeing models and their weights written separately. This is not the case when we use the models from streaming assets, so any help concerning memory management concerning models and Barracuda would be appreciated. 

On the other hand, the moment we load our models and start execution our RAM consumption jumps to 1.3GB of used RAM while total reserved RAM goes to 2GB. We had pinpointed the problem to encoder(this is our first model) output where we think the results is kept in RAM memory before it is passed to decoder model. Both of our models run on GPU(using ComputePrecomplied backend) so we wish to know if there is some solution that could bypass writing to RAM with the help of ComputeShaders or anything alike.

Our main issue is that we are unable to run the application on most of android devices because of the enormous RAM usage. You can find the models in the link below. Thank you all.

https://we.tl/t-GBZ1QBVrIL

![Screenshot_1](https://user-images.githubusercontent.com/79917924/117628367-5c24b200-b179-11eb-89fb-d0edb7ad1d19.jpeg)

**Comment by AlexRibard** [comment]
@CoaNewco thank you for rising the issue. I'll take a look at your models and investigate a bit.
To be clear you are asking about CPU RAM use? Is GPU RAM usage acceptable in your case?

In the meanwhile let me clarify a bit:
* at import (from ONNX), spike in RAM use is to be expected as we are performing model optimizations. These should be going away when running the application. In this case or when loading a .nn model, some temporary spike is to be expected as we need to load the weights and model info from CPU. But yes, once it's on the GPU it should go away.
* during execution, we need more RAM than only the weights. The weights are stored in RAM to not need to re-alloc them every frame. And we need a little more to store intermediate results. Do note we do dispose of intermediate results when not needed.

But given that your models are around 358MB on disk and you see a 2GB spike it is worth investigating.
I'll let you know if I find something

**Comment by CoaNewco** [comment]
Hello @AlexRibard :) I am asking about CPU RAM. We have been running some tests and doing some troubleshooting concerning RAM consumption and we've noticed some issues. First of all, when we reference our models in the inspector, automatically we see 270MB + 88MB writing to RAM. Now, this data is type of byte[] and it stays in RAM even though we unreference our models after loading them.

Next RAM spike is connected to weights and it occurs when we actually load the models, you've addressed this in the post above. We understand this is necessary, but unfortunately we have some 90,000,000 weights in both models. 

The BIGGEST concern is data transition between two models. We use encoder output as input tensor for our decoder model and immediately after calling encoder.PeekOutput() we see huge RAM ramp-up - anything between 600-1000MB, depending on encoder input dimensions. 

I've just wanted to give you an update which might help you help me in return :D Once again, thanks for all of your help and support!

**Comment by AlexRibard** [comment]
I took a look at things on our end.
There was indeed some gc-allocation that could be avoided. So I optimized that a bit (I'll push it for the next version)
Here is what I get overall:
* loading encoder: 0.6GB editor 0.53GB mobile
* running encoder: 0.69GB editor mobile : too slow
* loading decoder: 254.1MB editor 177MB mobile
* running decoder: 304.1MB editor mobile : too slow

* loading both models: 0.83GB editor 0.75GB mobile

* running both models: 0.87GB editor, mobile (too slow)

Do note the difference in RAM use on mobile vs editor. In editor we need to load the model for the inspector so that explains the spike there.

Your two models are respectively 88.3MB and 270MB so I'd say the 780MB on mobile is reasonable.

Here is a few tips:
* you can try using Compute worker instead of ComputePrecompiled, it'll be a bit slower but we do some extra optimizations that cost gc alloc
* if you alloc a tensor from the CPU and then feed that to the GPU it will create a gc alloc due to storing the data in CPU first
* when ping-ponging between models, don't read from the Tensor that will create a CPU copy. Just feed the PeekOutput to the next model (everything will be in GPU so no more alloc)

But I think what is holding you back from running on mobile isn't the RAM use but more the actual performance of the model.
The two models take around 1.2s on my GTX1080 to run so it's definitely too slow for mobile.
Try to reduce a bit the number of channels.

Hope that helps :)

P.S: I was not able to notice the huge spike when transitioning between models... so there might be something going on here...
So I'd look into that.
On my end I am just doing 
```
m_worker.Execute(m_input);
m_worker2.Execute(m_worker.PeekOutput());
```
And didn't notice the spike. So do double check you are not reading from the Tensor, forcing a CPU copy (and thus creating a 50MB spike). But 600MB is a bit crazy. Do email us at barracuda-support@unity3d.com if you have a repro case

**Comment by CoaNewco** [comment]
```
   public class OurClass { 
    // Lots of member variables ... 

    private Tensor EncoderTensor;

    private NNModel encoderModel;
    private NNModel decoderModel;

    // even more member variables
    private void Start() {
        //variable initialisation and other logic
    }

    //some functions and other logic here
    private void ModelLoading() {
        IWorker WorkerEncoder = ModelLoader.Load(encoderModel).CreateWorker(WorkerFactory.Device.GPU);
        IWorker WorkerDecoder = ModelLoader.Load(decoderModel).CreateWorker(WorkerFactory.Device.GPU);
    }

    private IEnumerator ModelsExecution() {    //this is how we handle model execution
        
        EncoderTensor = TransformInput();

        WorkerEncoder.Execute(EncoderTensor);    // --- HERE IS WHERE THE SPIKE IN RAM ACTUALLY HAPPENS

        yield return new WaitForSeconds(3);  // used for the purpose of testing


        WorkerDecoder.Execute(WorkerEncoder.PeekOutput("output"));
        Tensor argmaxOutput = WorkerDecoder.PeekOutput("output_argmax");

        EncoderTensor.Dispose();
        argmaxOutput.Dispose();
    }


    public Tensor TransformInput() {

        originalImageTexture = TextureTools.scaled(originalImageTexture, ENCODER_IMAGE_WIDTH, ENCODER_IMAGE_HEIGHT);
        Color32[] pix = originalImageTexture.GetPixelData<Color32>(0).ToArray<Color32>();
        ConvertPixels(pix);

        float[] floatValues = new float[pix.Length * 3];
        for (int i = 0; i < pix.Length; ++i)
        {
            Color color = pix[i]; // width*height*channels  -  pixel count

            floatValues[i * 3 + 0] = (color.r - Mean[0]) / ImageSTD[0];
            floatValues[i * 3 + 1] = (color.g - Mean[1]) / ImageSTD[1];
            floatValues[i * 3 + 2] = (color.b - Mean[2]) / ImageSTD[2];
        }
        
        return new Tensor(1, ENCODER_IMAGE_HEIGHT, ENCODER_IMAGE_WIDTH, 3, floatValues);
    }

}
```

This would be the outline of the code that we've found to be 'problematic'. It's after WorkerEncoder.Execute(EncoderTensor); function where writing to RAM happens. Could _combining_ of our models into one **single** model possibly bypass this? 

And I wish to address one more point concerning inference time. You've mentioned that it takes 1.2 seconds for models to finish with execution, but our data is completely different. When we look at ModelsExecution() in profiler we see that the everything is done in less than 2 milliseconds NOT seconds. Are we looking at this wrong? Because 1.2 SECONDS for execution is definitely something we can't live with.

**Comment by AlexRibard** [comment]
* inference time, yeah that could be explained by the input size you are using vs the one I tried.
If the timings are fine on your end then things are good :) (btw on latest barracuda you have optimized softmax, let me know how fast it is now)

* the problem of your code is here: `EncoderTensor = TransformInput();`
You are first of all downloading a texture to CPU, modifying it, copying that into the tensor and then uploading this to the GPU.
That create a lot of gc alloc.
This will be improved a bit by my upcoming optimizations, but I'd say the main issue is here.
Are you able to keep all that normalization code on the GPU? Are you comfortable doing this in a compute shader?
We haven't exposed much in terms of scaling from going from TextureToTensor, but the other way you can.
So I'd suggest doing the following:
```
var tmp = new Tensor(srcTexture:inputRenderTexture, channels:3);
tmp.ToRenderTexture(target:tmpRenderTexture, batch:0, fromChannel:3, scale:1/ImageSTD, bias: -Mean/ImageSTD);

return new Tensor(tmpRenderTexture,3)
```
It is a bit inefficient since we create two tensors, but everything will be kept on the GPU.
Alternatively you could add that normalization code (x - Mean) / ImageSTD; at the beginning of the encoder model. You can even add the texture scaling with a `Resize` layer.
That would be the most optimal.

Let me know

**Comment by CoaNewco** [comment]
Hi, we've managed to work some things out through shaders. But of course, another problem manifested itself. We've noticed that Barracuda converted two kernels in our conv2d layers from 3x3x512x512 to 9x9x512x512. Why did this happen? Because of this problem there is around 36 million weights that are inside the model now(all of them are zeros), and that is causing RAM problems, among some other stuff. I'll keep you posted of course, but I would like to know how to solve kernel problem. Thank you. 

P.S. Is there any way we can transfer execution to AI cores on mobile phone SoC, instead of using GPU or CPU?

**Comment by AlexRibard** [comment]
* We have a optimization pass at import to fuse consecutive linear layers when it is efficient (ie produce a speedup).
In your case, fusing those two conv would provide in runtime speedup.
There is no real reason why they should be 0, unless there already was 0 presents in the original weights.
But you can always remove optimizations in the model inspector: `OptimizeModel`
Looking at our code, I think that flag is sadly broken :/
I'll go ahead and fix than, in the meantime if the checkbox does not work, you can comment out 
`IntermediateToRunnableNHWCPass.cs > linearLayerFusingPass.Run(ref model);`
Also let me know how much we actually win with this optimization pass in your case.

* At the moment, sadly you cannot take advantages of AI cores. We do plan on supporting it, but atm it is in development.

**Comment by CoaNewco** [comment]
There's this 
The package cache was invalidated and rebuilt because the following immutable asset(s) were unexpectedly altered:
  Packages/com.unity.barracuda/Barracuda/Runtime/Core/Compiler/Passes/IntermediateToRunnableNHWCPass.cs
  so I can't actually change it. And the flag isn't working also :S

**Comment by AlexRibard** [comment]
I think for this you'll need to download the github source and link to that locally.
I'll see if we can have a new version soon to unlock you

**Comment by CoaNewco** [comment]
Hello again :) Little update on the whole deal. First, by fusing layers we get around 36000000 more weights, and they're all zeros. This way we get a bottleneck
![milis320](https://user-images.githubusercontent.com/79917924/118399366-1e35fb00-b65d-11eb-8107-4f521435aa19.png)

 on those two problematic Conv2D layers. I've tried deleting layer weights by accessing them via DataSet[0].shape which I can manage, but I couldn't actually figure out how to write actual weights without zeros back. I've also tried removing whole layer, but once again, I have offset problem. 

I've tried linking package from github locally and commenting out the lines you've pointed me to, but again, nothing changed. Could you point me to any working solution? Because, having things as they are, we can't move forward with our model. :/

**Comment by AlexRibard** [comment]
Once you have downloaded the github and commented the lines out and correctly linked the local project, you need to re-import the onnx model. Have you done this?
If this doesn't work, a quick hack you can do is add a `LeakyRelu` with alpha = 1 (ie Identity layer, but won't get optimized away) between your two conv layers. This should stop the Conv being fused.

**Comment by CoaNewco** [comment]
I've done everything you have suggested, but no luck. Everything still stays the same, still I have 38 something million weights added. Is there any way I can remove unwanted weights from the layer and then add correct float array?

[Here ](https://we.tl/t-fN6e3DC3cc)is the model that we are using right now. I would be grateful if you could help me around this particular issue.

**Comment by AlexRibard** [comment]
Ok, I took a look at the model.
Sorry about the misleading. There is indeed no Conv fusing. The problem is coming from the `dilatations` (cf Conv_119)
unfortunately our current implementation for `dilatations` dilates the conv kernel and adds 0. 
This is indeed inefficient, but we didn't have a use case for that until now.
I'll create a task for us (it shouldn't take very long) and let you know when it lands.

In the meantime, there isn't much workaround you can do unless you remove them...
Maybe you can replace them with `ConvTranspose`, as the operator is somewhat similar?

**Comment by CoaNewco** [comment]
Hello from me again! First, you're right, kernel dilations are the culprit of all the zero weights, and unfortunately we had to give way to other solutions due to unresolved dilations implementation. We have managed to remove the unnecessary weights(zeroes) from the conv layers, but in doing so we have compromised the model's structure and also we lost the dilations' purpose. Do you have any estimate when will the implementation for dilations be complete? In this [link](https://we.tl/t-YxgRL7t628) you can find the manually altered model which we use. 

Being unable to achieve anything of value with the model above we have turned to another model and yet another idea :) We are now focused on MobileNetV2 pretrained for semantic segmentation use case. We were absolutely amazed with performance on PC, both the speed and the precision. The problem arose when we tried to execute the model on Android device(S20+). The app instantly crashes when we execute the model. This shocked us because said model has 44 something million weights less than encoder-decoder(combined model). Also, MobileNetV2 uses DepthwiseConv2D layer, which is faster than Conv2D layers. The problem lies in RAM _**overload**_ in the exact moment of model's execution. Everything happens in the same frame and System Used Memory jumps from 0.51 GB to 3.14 GB  instantly, which causes app to crash on android. We've went profiling and found that the usual suspects are guilty again, but this time they went to far. GenericWorker.Execute() and then propagated lower - SetInput() took nearly one gb of ram as well as 1265ms while while Execute() took around half a gigabyte and 1052 millis. The main actor in my personal frustration and pain is the fact that this model was supposed to work faster than the one above. Do you have any advice at all on how to fix this? [This](https://we.tl/t-sqa7YRCelz) is the model we're using.

Publishing deadline is coming in less than two months, and we would prefer to stick with Unity and Barracuda for NN inference so we sincerely hope that some big update is around the corner. Every day we're hoping like a kid on Christmas day. :D One day, hopefully soon the gift is going to come. 

Thank you for all the advice from me personally, but also big thanks from the whole team. Cheers!

**Comment by AlexRibard** [comment]
@CoaNewco I'll check your latest model and see if I can spot anything wrong.
Again if you are passing your input from the CPU to GPU that would cause some gc alloc.
If you could keep everything on the GPU that would be great.
If like previously you are only doing image rescaling try using the Tensor to Texture tricks I mentioned a while back

The team would be intresting in knowing more about your project and deadlines. Would you be ok with scheduling a meeting with us?
You can email us at barracuda-support@unity3d.com
Let me know

**Comment by CoaNewco** [comment]
I think I'm pushing everything to GPU, but I'll double and triple check that of course. We're using TensorToTexture, and that was a real life saver, thanks for that. :) Also all of the preprocessing is done on GPU using shaders, so I found myself dumbstruck when this happened. 

I would, definitely. As would the rest of the team. My colleague Vladimir, who is also our R&D director, will contact you soon. 
Again, thanks for all the help. :)

**Comment by AlexRibard** [comment]
@CoaNewco I tested your new model on master (with the gc-alloc optims I mentioned). Happy to say that it runs without crashing.
It uses 21MB Mono and 495.2MB Gfx.
I tested on the latest public version, there it crashes.
So I'm hopeful that it'll work when you get the promised optimizations

**Comment by coaodje** [comment]
Hi! Do you know when will the GC allocation issue be solved?

**Comment by AlexRibard** [comment]
Fix has landed on master, it will be present in an upcoming version.
Working on a back-port to 1.4 at the moment
